test/Microsoft.AspNetCore.Server.KestrelTests/BadHttpRequestTests.cs

```diff
@@ -1,72 +1,119 @@
+        [InlineData("GET / HTTP/1.1 \r")]
+        [InlineData("GET / HTTP/1.1a\r")]
+        [InlineData("GET / HTTP/1.0\n\r")]
+        [InlineData("GET / HTTP/1.2\r")]
+        [InlineData("GET / HTTP/3.0\r")]
+        [InlineData("GET / H\r")]
+        [InlineData("GET / HTTP/1.\r")]
+        [InlineData("GET / hello\r")]
+        [InlineData("GET / 8charact\r")]
+        // Missing LF
+        [InlineData("GET / HTTP/1.0\rA")]
+        public async Task TestBadRequestLines(string request)
         {
             using (var server = new TestServer(context => { return Task.FromResult(0); }))
             {
                 using (var connection = new TestConnection(server.Port))
                 {
-                    var receiveTask = Task.Run(async () =>
-                    {
-                        await connection.Receive(
-                            "HTTP/1.1 400 Bad Request",
-                            "");
-                        await connection.ReceiveStartsWith("Date: ");
-                        await connection.ReceiveForcedEnd(
-                            "Content-Length: 0",
-                            "Server: Kestrel",
-                            "",
-                            "");
-                    });
-
-                    try
-                    {
-                        await connection.SendEnd(request).ConfigureAwait(false);
-                    }
-                    catch (Exception)
-                    {
-                        // TestConnection.SendEnd will start throwing while sending characters
-                        // in cases where the server rejects the request as soon as it
-                        // determines the request line is malformed, even though there
-                        // are more characters following.
-                    }
+                    await connection.SendEnd(request);
+                    await ReceiveBadRequestResponse(connection, request == "GET / HTTP/1.1\r");
+                }
+            }
+        }
 
-                    await receiveTask;
+        [Theory]
+        [InlineData(" ")]
+        [InlineData("GET  ")]
+        [InlineData("GET / HTTP/1.2\r")]
+        [InlineData("GET / HTTP/1.0\rA")]
+        public async Task ServerClosesConnectionAsSoonAsBadRequestLineIsDetected(string request)
+        {
+            using (var server = new TestServer(context => { return Task.FromResult(0); }))
+            {
+                using (var connection = new TestConnection(server.Port))
+                {
+                    await connection.Send(request);
+                    await ReceiveBadRequestResponse(connection, false);
                 }
             }
         }
+
+        private async Task ReceiveBadRequestResponse(TestConnection connection, bool expectedConnectionCloseHeader)
+        {
+            await connection.Receive(
+                "HTTP/1.1 400 Bad Request",
+                "");
+
+            if (expectedConnectionCloseHeader)
```

Comment:
According to https://tools.ietf.org/html/rfc7230#section-6.6:

> The Connection header field (Section 6.1) provides a "close" connection option that a sender SHOULD send when it wishes to close the connection after the current request/response pair.

So if the server is closing the connection on an error response, it should set `Connection: close`.

A related question is whether the server should close the connection on an error response, or reuse the connection for subsequent requests.  I couldn't find anything in RFC 7230, but https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.1.2 seems to indicate the server should keep the connection open:

> A significant difference between HTTP/1.1 and earlier versions of HTTP is that persistent connections are the default behavior of any HTTP connection. That is, unless otherwise indicated, the client SHOULD assume that the server will maintain a persistent connection, even after error responses from the server.

It might be reasonable to have different behavior for different errors.  For example, an invalid request could close the connection, while something like `404 Not Found` could keep the connection alive.